### DISCUSSION TITLE
Enhance map with custom decorators

### DISCUSSION THREAD

**Discussion by GalStan** [MichalLytek/typegraphql-prisma#210] [Q&A] [answered]
Does resolverEnhanceMap works with custom decorators that change original descriptors? I'm trying to wrap original resolver function but seems it doesn't work.

**Comment by MichalLytek** [comment, marked as answer]
Hard to say, `enhance` helpers are just calling the decorator functions in a way that TS does in the transpiled code.

I encourage you to debug and see why it could not be working 🤔

**Reply by GalStan** [reply]
I've tested my custom decorator setting explicitly on resolver and everything works fine. But when I try to use enhance map for generated one it is being execute bot descriptor seems not to be changed.

**Reply by GalStan** [reply]
I think the problem is that mutation of result from getOwnPropertyDescriptor has no effect on original descriptor https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/getOwnPropertyDescriptor

` for (const allActionsDecorator of allActionsDecorators) {
          allActionsDecorator(
            crudTarget,
            resolverActionName,
            Object.getOwnPropertyDescriptor(crudTarget, resolverActionName)!,
          );
          allActionsDecorator(
            actionTarget,
            resolverActionName,
            Object.getOwnPropertyDescriptor(actionTarget, resolverActionName)!,
          );
        }
      }`

**Reply by MichalLytek** [reply]
This is what TypeScript emits:
```js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
```

So if I understand correctly, you're leveraging the mechanism described in TS docs:
> If the method decorator returns a value, it will be used as the Property Descriptor for the method.

And my approach of calling the decorators is incorrect because I don't collect the returned value nor pass it to next decorators, as well as not call the `Object.defineProperty` to apply the descriptor?

**Reply by GalStan** [reply]
According to TS docs you can just mutate original descriptor without returning it:
```javascript
function enumerable(value: boolean) {
  return function (target: any, propertyKey: string, descriptor: PropertyDescriptor) {
    descriptor.enumerable = value;
  };
}
```
But if you could add the ability to return descriptor and use it on resolvers, that would be great.

**Reply by MichalLytek** [reply]
I think we need both and `defineProperty`, so it would be the best to just use the `tslib` helpers:
```js
(0, tslib_1.__decorate)([
    TypeGraphQL.Field(_type => CreatorWhereInput_1.CreatorWhereInput, {
        nullable: true
    }),
    (0, tslib_1.__metadata)("design:type", CreatorWhereInput_1.CreatorWhereInput)
], ProblemLikedByArgs.prototype, "where", void 0);
```

**Reply by GalStan** [reply]
I think this is the best solution

**Reply by MichalLytek** [reply]
Released on `v0.18.0` 🎉 
Let me know if that fixes your issue.

**Reply by GalStan** [reply]
Hm, seems it is broken now. I'me getting `undefined` as a descriptor

**Reply by GalStan** [reply]
Think you should pass `null` rather than `void` here 
```javascript
tslib.__decorate(allActionsDecorators, crudTarget, resolverActionName, void 0);
```
cause now this helper passes void (undefined) as a descriptor 
```javascript
 __decorate = function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
```

**Reply by MichalLytek** [reply]
You're right! Fixed that in db987c1, released in `v0.18.1` 💪

**Reply by dsmurl** [reply]
Are there docs or direction for users wanting to make custom decorators?
